### PR TITLE
Fix behavior keepalive

### PR DESCRIPTION
**Pull request by maxvalentini77**
### I found three issues:

1: In some cases, call an invalid URL, resulting in a response with code 404. (still refresh the session)
2: Update the article hits counter.
3: Work in different ways with "Search Engine Friendly URLs" enabled or disabled.
### How to reproduce number 1
- Enable "Search Engine Friendly URLs"
- Set "Session Lifetime" to 1 (not required but useful to test faster!)
- Create a menu of type "Menu Heading" and call it "heading"
- Create a menu that point to an article and add it as child of "heading"
- Create a Login module and put it some where in the page (add the "behavior.keepalive")
- Use the article menu (you will be redirected to a page with an URL like "www.host.net/index.php/heading/article")
- Wait some "Session Lifetime" minutes and an ajax call to the URL "www.host.net/index.php/heading/" will be executed but ends with a 404 error

The session is refreshed but it isn't nice to generate a 404 error that, for example, ends up in the apache logs.
### How to reproduce number 2
- Set "Session Lifetime" to 1 (not required but useful to test faster!)
- Create a menu that point to an article and set it as default
- Create a Login module and put it some where in the page (add the "behavior.keepalive")
- Go to the homepage of the site, the article will be displayed with a counter hits = C
- Wait some "Session Lifetime" minutes and an ajax call to the homepage will be executed (increasing the hits counter)
- Refresh the page

The counter was increased by 1+X where X is the number of ajax calls.
### How to reproduce number 3
- Disable "Search Engine Friendly URLs"
- Set "Session Lifetime" to 1 (not required but useful to test faster!)
- Create a menu of type "Menu Heading" and call it "heading"
- Create a menu that point to an article and add it as child of "heading"
- Create a Login module and put it some where in the page (add the "behavior.keepalive")
- Use the article menu (you will be redirected to a page with an URL like "www.host.net/index.php?option=com_content&view=article&id=12&Itemid=160")
- Wait some "Session Lifetime" minutes and an ajax call to the URL "www.host.net/" will be executed.

This is the same procedure to reproduce number 1 but ends with a different behavior that, if the homepage display an article, generate the same issue of number 2.
(This because the homepage of the site has been called.)
### Proposed solution

I modified the URL of the ajax call from "./" to:

``` php
JRoute::_("index.php?option=com_ajax&format=json", false)
```

This call refresh the session without generate any html and/or call model methods that can lead to unexpected/unwanted behaviors (for example in an increased article hits counter).
This call is available on both site and administrator applications.

To test it in the administrator application (for example) go to the edit category page that include the "behavior.keepalive" and wait for the ajax call.

### CODE REVIEW

**Comment by Fedik** [comment]
thanks @maxvalentini77 I had similar idea, and found this is already suggested :smile: 

couple note:
 you do not need to call `JRoute` as it useless here, enough:
 `JUri::base(true) . 'index.php?option=com_ajax&format=json';`

 and can you please make `$refreshTime` fixed, example 1 minute , so we could solve also that issue #6730

something like:
```php
JFactory::getDocument()->addScriptDeclaration('
!window.setInterval(function(){
var r; try{ r = window.XMLHttpRequest ? new XMLHttpRequest() : new ActiveXObject("Microsoft.XMLHTTP")}catch(e){};if(r){r.open("GET","'.$url.'",true);r.send(null)}
}, 60000);
');
```

**Comment by maxvalentini77** [comment]
Hi @Fedik
I made the changes.

Thank you

**Comment by Fedik** [comment]
I am sorry, I missed slash in my example, for url:
`JUri::base(true) . '/index.php?option=com_ajax&format=json';`

**Comment by maxvalentini77** [comment]
Added the slash and tested, seems to work well.
Time is now fixed to 60 seconds, I don't know if also solves the #6730.

**Comment by Fedik** [comment]
**test** works for me good, also fix #6730 issue for me
thanks!

**Comment by Fedik** [comment]
**test**
still works good :smile:

**Comment by Fedik** [comment]
it is ready to "RTC" or? :smiley: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/4904">issues.joomla.org/joomla-cms/4904</a>.</sub>

**Comment by demis-palma** [comment]
@test successful
@Fedik I guess that your test has not been notified because it lacks the (at) symbol.

**Comment by Fedik** [comment]
the (at) symbol useless here :wink: 
people use it as quick way to make "test" word **bold** :smile:

**Comment by zero-24** [comment]
Thanks for the bump and the tests @Fedik and @demis-palma

RTC'ing now!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/4904">issues.joomla.org/joomla-cms/4904</a>.</sub>

File: libraries/cms/html/behavior.php [hunk 1]
```diff
@@ -599,28 +599,18 @@ public static function keepalive()
...
-		$document = JFactory::getDocument();
 		$script = 'window.setInterval(function(){';
 		$script .= 'var r;';
 		$script .= 'try{';
 		$script .= 'r=window.XMLHttpRequest?new XMLHttpRequest():new ActiveXObject("Microsoft.XMLHTTP")';
 		$script .= '}catch(e){}';
-		$script .= 'if(r){r.open("GET","./",true);r.send(null)}';
-		$script .= '},' . $refreshTime . ');';
+		$script .= 'if(r){r.open("GET","' . $url . '",true);r.send(null)}';
+		$script .= '},60000);';
```
Comment (dgt41) [inline libraries/cms/html/behavior.php]: Shouldn’t we repeat the call every X seconds? Where X is the option set in global config. So all I am saying is: restore this line here `$script .= '},' . $refreshTime . ');';` and a few lines above `$lifetime = ($config->get('lifetime') * 60000);`

Right now this is fixed to a call per minute, which for some busy sites will create more load...

Comment (maxvalentini77) [inline libraries/cms/html/behavior.php]: In the begging i don't change the time, but @Fedik open the issue #6730 and ask to me to reduce the time to 1 minute to fix also the #6730.
With this change the request is very "light" do not generate any html reducing time and size of the request (without this change can request the homepage!).

Comment (dgt41) [inline libraries/cms/html/behavior.php]: @maxvalentini77 I wasn’t aware of #6730…

**Comment by maxvalentini77** [comment]
I notice that the session not always be refreshed if the session lifetime is set to 1 minute.
@Fedik and @demis-palma can check again that continue to works well to you too?
(sorry for the inconveniente)

I also noticed that can be set a zero or negative value in the configuration page.
Zero behavior is equals to default of 15 minutes.
Negative value cause the impossibility to use the application!

**Comment by Fedik** [comment]
@maxvalentini77 I already use this on my sites, where default session handler is File or APC, 
and refresh session each minute works good, if compare with original

session lifetime 1 minute I think have not much sense, not sure that we need to worry about it :wink:
but maybe someone has better arguments :smile:

**Comment by Kubik-Rubik** [comment]
Thank you @maxvalentini77! Merged.

**Pull request closed without merging**